src/main/scala/edu/berkeley/cs/amplab/mlmatrix/RowPartitionedMatrix.scala

```diff
@@ -311,6 +314,80 @@ object RowPartitionedMatrix {
     data
   }
 
+  // Convert a DataFrame of all DoubleType columns to a RowPartionedMatix
+  def fromDataFrame(df: DataFrame): RowPartitionedMatrix = {
+    require(df.dtypes.forall(_._2 == "DoubleType"),
+      "The provided DataFrame must contain all 'DoubleType' columns")
+    fromMatrix(dataFrameToMatrix(df))
+  }
+
+  def fromDataFrame(
+      df: DataFrame,
+      rowsPerPartition: Seq[Int],
+      cols: Int): RowPartitionedMatrix = {
+    require(df.dtypes.forall(_._2 == "DoubleType"),
+      "The provided DataFrame must contain all 'DoubleType' columns")
+    new RowPartitionedMatrix(
+      dataFrameToMatrix(df, rowsPerPartition, cols).map(mat => RowPartition(mat)),
+        Some(rowsPerPartition.sum), Some(cols))
+  }
+
+  // Convert a DataFrame of all DoubleType columns to a RDD[DenseMatrix[Double]]
+  def dataFrameToMatrix(
+      df: DataFrame,
+      rowsPerPartition: Seq[Int],
+      cols: Int) = {
+    val matrixRDD = df.map(x => x.toSeq.toArray).map(y => y.map(z => z.asInstanceOf[Double]))
+    val rBroadcast = matrixRDD.context.broadcast(rowsPerPartition)
+    val data = matrixRDD.mapPartitionsWithIndex { case (part, iter) =>
+      val rows = rBroadcast.value(part)
+      val matData = new Array[Double](rows * cols)
+      var nRow = 0
+      while (iter.hasNext) {
+        val arr = iter.next()
+        var idx = 0
+        while (idx < arr.size) {
+          matData(nRow + idx * rows) = arr(idx)
+          idx = idx + 1
+        }
+        nRow += 1
+      }
+      Iterator(new DenseMatrix[Double](rows, cols, matData.toArray))
+    }
+    data
+  }
+
+  def dataFrameToMatrix(df: DataFrame): RDD[DenseMatrix[Double]] = {
+    val matrixRDD = df.map(x => x.toSeq.toArray).map(y => y.map(z => z.asInstanceOf[Double]))
```

Comment:
If I'm not wrong `matrixRDD` here is a `RDD[Array[Double]]` ? If so can we just call `arrayToMatrix(matrixRDD)` here ?